libs/ml/src/core/node.cpp

```diff
@@ -0,0 +1,236 @@
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2020 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "ml/core/node.hpp"
+#include "ml/core/subgraph.hpp"
+
+namespace fetch {
+namespace ml {
+
+using Shape       = fetch::math::SizeVector;
+using ShapeVector = std::vector<fetch::math::SizeVector>;
+
+/**
+ * returns the stored operation type and syncs it with operation type of
+ * underlying Ops.
+ * @return
+ */
+template <typename TensorType>
+OpType Node<TensorType>::OperationType()
+{
+  if (operation_type_ != op_ptr_->OperationType())
+  {
+    FETCH_LOG_ERROR(this->name_.c_str(), "Node operation type (" +
+                                             std::to_string(static_cast<int>(operation_type_)) +
+                                             ") and underlying Ops operation code (" +
+                                             std::string(op_ptr_->Descriptor()) + ") mismatch!");
+    operation_type_ = op_ptr_->OperationType();
+  }
+  return operation_type_;
+}
+
+template <typename TensorType>
+bool Node<TensorType>::HasValidCache()
+{
+  return static_cast<bool>(cached_output_status_ == CachedOutputState::VALID_CACHE);
+}
+
+template <typename TensorType>
+void Node<TensorType>::SetBatchOutputShape(const Shape &new_shape)
+{
+  op_ptr_->SetBatchOutputShape(new_shape);
+}
+
+template <typename TensorType>
+void Node<TensorType>::SetBatchInputShapes(const ShapeVector &new_shapes)
+{
+  op_ptr_->SetBatchInputShapes(new_shapes);
+}
+
+template <typename TensorType>
+const ShapeVector &Node<TensorType>::BatchInputShapes()
+{
+  return op_ptr_->BatchInputShapes();
+}
+
+/**
+ * @brief BatchOutputShape computes an output shape of the Node, if only 1 data slice (e.g.
+ * with batch size == 1) is provided to the Graph input. If there is no cached output shape,
+ * the method is recursively called until either a cached shape or input Node is encountered.
+ * @return vector of SizeType.
+ */
+template <typename TensorType>
+Shape Node<TensorType>::BatchOutputShape()
+{
+  Shape const &candidate = op_ptr_->BatchOutputShape();
+  bool const   op_is_subgraph =
```

Comment:
this is an unnecessary step. you could just:

auto subgraph_ptr = dynamic_pointer_cast<Subgraph>(op_ptr_);

if (!candidate.empty() && subgraph_ptr)
{

}

Reply:
if it's a matter of clarity I would say that the comment on line 84 handles that for you - "if the op is not a subgraph ..."

Reply:
agreed